### PR TITLE
Graph path planning gather resources

### PR DESCRIPTION
**Pull request by SanderBouwman**
- [x] Code review
- [x] Comments
- [ ] Functional tests written
- [ ] Functional tests checked by someone other than the author(s)
- [x] No linting errors
- [x] Unit tests passes
- [x] Technical documentation written
- [ ] Technical documentation reviewed

### CODE REVIEW

File: docs/software_architecture/chapters/path-planning.tex [hunk 1]
```diff
@@ -0,0 +1,25 @@
+\section{Path Planning}
+\label{sec:pathplanning}
+For path planning we’ve chosen to use A*. In the following chapter we dive further into details how we use A*.
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/path-planning.tex]: Dijkstra's algorithm with the A* optimalization

File: docs/software_architecture/chapters/path-planning.tex [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+We’ve chosen for the Manhattan heuristic. This heuristic is easy to implement and works pretty well for our game. We are using the following weights for determining the step cost:
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/path-planning.tex]: We've chosen to use the Manhattan heuristic

Comment (JeroenVinke) [inline docs/software_architecture/chapters/path-planning.tex]: why? design decision

File: docs/software_architecture/chapters/path-planning.tex (hunk 1 above)
Comment (JeroenVinke) [inline docs/software_architecture/chapters/path-planning.tex]: why? design decision

File: docs/software_architecture/chapters/behaviour.tex [hunk 3]
```diff
@@ -0,0 +1,12 @@
+\section{Behaviour}
+\label{sec:behaviour}
+We've chosen to use the goal driven behaviour approach. Since a goal can consist of multiple smaller goals, the Composite Pattern is a good solution to this problem. You can have small goals such as 'TraverseEdge' and also bigger goals like 'Work' and still treat them the
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/behaviour.tex]: why goal driven behavior? design decision

File: docs/software_architecture/chapters/composite.tex [hunk 4]
```diff
@@ -0,0 +1,27 @@
+\subsection{Composite goals}
+We created a couple of Composite goals for moving entities. There is a class diagram \cref{fig:goalcomposite-inherit} in the appendix that illustrates the structure. This diagram doesn't show all composite goals, however is should give an impressions how we implemented this.
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/composite.tex]: We created a couple of Composite goals for moving entities **which we describe below**.

File: docs/software_architecture/chapters/composite.tex [hunk 5]
```diff
@@ -0,0 +1,27 @@
+\subsection{Composite goals}
+We created a couple of Composite goals for moving entities. There is a class diagram \cref{fig:goalcomposite-inherit} in the appendix that illustrates the structure. This diagram doesn't show all composite goals, however is should give an impressions how we implemented this. 
+
+\subsubsection{Think}
+The Think goal is a goal that never gets removed. This goal is needed 
+to determine an entity's next goal and activates that goal. It does so by 
+calling the goals evaluator class, which returns a desirability value. The
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/composite.tex]: why use evaluators? design decision + source would be nice here

File: docs/software_architecture/chapters/atomic.tex [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+The obstacle avoidance goal activates the obstacle avoidance steering 
+behaviour. When this goal is activated, it adds the steering behaviour to the 
+entity's behaviours. Once the distance to the target that it needs to avoid 
+is big enough, the goal is completed and the steering behaviour also gets 
+removed from the entity.
+
+\subsubsection{Drop Resources}
+\label{sec:dropresources}
+Once an entity has gathered enough resources, it needs to drop them at a 
+warehouse/depot. This goal simply removes the resources the entity gathered and adds them to the resources of the player. If it dropped all of it's resources at the warehouse, the goal
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/atomic.tex]: **When** it has dropped all of it's resources

File: docs/software_architecture/chapters/atomic.tex [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+\subsubsection{Drop Resources}
+\label{sec:dropresources}
+Once an entity has gathered enough resources, it needs to drop them at a 
+warehouse/depot. This goal simply removes the resources the entity gathered and adds them to the resources of the player. If it dropped all of it's resources at the warehouse, the goal 
+is completed.
+
+\subsubsection{Gather Resource}
+\label{sec:gatherresource}
+This goal is calls the Gather() method from a resource entity. This method extracts resources from the resource entity and adds it to the entity that is gathering the resource. Once the resource entity has been depleted or the maximum carrying capacity of the gathering entity has been reached this goal will be completed.
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/atomic.tex]: remove "is" from "This goal is calls the Gather()"

File: docs/software_architecture/chapters/atomic.tex [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+heuristic (\cref{sec:pathplanning}). Once the path has been generated, it gets set as the active path 
+for the given entity. This is the only task it needs to complete before 
+getting removed from the containing goal.
+
+\subsubsection{Traverse Edge}
+\label{sec:traverseedge}
+This goal uses the Seek behaviour explained in \cref{sec:seek-behaviour}. Once 
+this class gets instantiated it adds an instance of SeekStrategy to the 
+entity's behaviour. The only thing it needs to do while processing, is 
+checking whether the entity has this behaviour, and if so if it's close
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/atomic.tex]: is to check whether

File: docs/software_architecture/chapters/appendix.tex [hunk 9]
```diff
@@ -0,0 +1,11 @@
+\begin{figure}[!htb]
+    \centering
+    \includegraphics[scale=0.75]{res/Goal.jpg}
+    \caption{The Goal<T> class.}\label{fig:goal}
+\end{figure}
+
+\begin{figure}[!htb]
+    \centering
+    \includegraphics{res/GoalComposite-Inherit.jpg}
```
Comment (JeroenVinke) [inline docs/software_architecture/chapters/appendix.tex]: this image is huuuuge

img

File: docs/software_architecture/chapters/path-planning.tex [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+To calculate the optimal path there are 3 variables. \\
+\textbf{\textit{g}} = Step cost \\
+\textbf{\textit{h}} = Heuristic estimate \\ 
+\textbf{\textit{f}} = Calculated total  \textbf{\textit{(h + g)}}
+
+We start off with measuring \textbf{\textit{h}}. We do this by calculating the distance from the starting node to the end node. We then calculate \textbf{\textit{f}} for all neighbors, a neighbor is a node to which there is an edge from the current node. \textbf{\textit{f}} is calculated by adding the step cost\textbf{\textit{(g)}} from the starting node to the neighbor and the heuristic estimate\textbf{\textit{(h)}} of that neighbor. Once that process has been done for all neighbors we pick the neighbor with the lowest \textbf{\textit{f}}. The following process is then repeated, however the starting node is now the neighbor with the lowest \textbf{\textit{f}}. We are done when we’ve reached our end node.
+
+While calculating the path we also set parent nodes, this is used to build the path when we’ve reached the final node. The path is built by using a recursive method, which can be seen in the figure below. 
+
+\begin{figure}[!htb]
```
Comment (Rboin) [inline docs/software_architecture/chapters/path-planning.tex]: Would like to see this changed to a proper code listing instead of an image.

**Pull request merged**